### ISSUE TITLE
simple_bind() takes a long time

### ISSUE THREAD

**Issue by ufoym** [dmlc/mxnet#1557]
I've just tested _[Predict with Pre-trained model](https://github.com/dmlc/mxnet/blob/master/example/notebooks/predict-with-pretrained-model.ipynb)_, and found that the whole pipeline took about 45s to process a single image (env: i7 CPU / 32G RAM / Titan X / SSD / Ubuntu x64 / Python 2.7 x64 / CUDA 7.0 / cuDNN v3). More specifically, it seems that most time spends on simple_bind().

**Code**

    import mxnet as mx
    import logging
    import numpy as np
    from skimage import io, transform

    logger = logging.getLogger()
    logger.setLevel(logging.DEBUG)

    # Load the pre-trained model
    prefix = "Inception/Inception_BN"
    num_round = 39
    model = mx.model.FeedForward.load(prefix, num_round, ctx=mx.gpu(2), numpy_batch_size=1)

    # load mean image
    mean_img = mx.nd.load("Inception/mean_224.nd")["mean_img"]

    # load synset (text label)
    synset = [l.strip() for l in open('Inception/synset.txt').readlines()]

    def PreprocessImage(path, show_img=False):
        # load image
        img = io.imread(path)
        print("Original Image Shape: ", img.shape)
        # we crop image from center
        short_egde = min(img.shape[:2])
        yy = int((img.shape[0] - short_egde) / 2)
        xx = int((img.shape[1] - short_egde) / 2)
        crop_img = img[yy : yy + short_egde, xx : xx + short_egde]
        # resize to 224, 224
        resized_img = transform.resize(crop_img, (224, 224))
        # convert to numpy.ndarray
        sample = np.asarray(resized_img) * 256
        # swap axes to make image from (224, 224, 4) to (3, 224, 224)
        sample = np.swapaxes(sample, 0, 2)
        sample = np.swapaxes(sample, 1, 2)
        # sub mean
        normed_img = sample - mean_img.asnumpy()
        normed_img.resize(1, 3, 224, 224)
        return normed_img

    # Get preprocessed batch (single image batch)
    batch = PreprocessImage('./1.jpg', True)
    # Get prediction probability of 1000 classes from model
    prob = model.predict(batch)[0]

**profile predict()**

    Line #      Hits         Time  Per Hit   % Time  Line Contents
    ==============================================================
       573                                               @profile
       574                                               def predict(self, X, num_batch=None, return_data=False, reset=True):
       575                                                   """Run the prediction, always only use one device.
       576                                                   Parameters
       577                                                   ----------
       578                                                   X : mxnet.DataIter
       579                                                   num_batch : int or None
       580                                                       the number of batch to run. Go though all batches if None
       581                                                   Returns
       582                                                   -------
       583                                                   y : numpy.ndarray or a list of numpy.ndarray if the network has multiple outputs.
       584                                                       The predicted value of the output.
       585                                                   """
       586         1          185    185.0      0.0          X = self._init_iter(X, None, is_train=False)
       587
       588         1            1      1.0      0.0          if reset:
       589         1            2      2.0      0.0              X.reset()
       590         1           16     16.0      0.0          data_shapes = X.provide_data
       591         2            2      1.0      0.0          data_names = [x[0] for x in data_shapes]
       592         1     39704804 39704804.0     99.7          self._init_predictor(data_shapes)
       593         1            2      2.0      0.0          batch_size = X.batch_size
       594         2            5      2.5      0.0          data_arrays = [self._pred_exec.arg_dict[name] for name in data_names]
       595         2            5      2.5      0.0          output_list = [[] for _ in range(len(self._pred_exec.outputs))]
       596         1            2      2.0      0.0          if return_data:
       597                                                       data_list = [[] for _ in X.provide_data]
       598                                                       label_list = [[] for _ in X.provide_label]
       599
       600         1            1      1.0      0.0          i = 0
       601         2          599    299.5      0.0          for batch in X:
       602         1            1      1.0      0.0              if num_batch is not None and i == num_batch:
       603                                                           break
       604         1            1      1.0      0.0              i += 1
       605
       606         1           32     32.0      0.0              _load_data(batch, data_arrays)
       607         1          127    127.0      0.0              self._pred_exec.forward(is_train=False)
       608         1            1      1.0      0.0              padded = batch.pad
       609         1            1      1.0      0.0              real_size = batch_size - padded
       610
       611         2            6      3.0      0.0              for o_list, o_nd in zip(output_list, self._pred_exec.outputs):
       612         1       102447 102447.0      0.3                  o_list.append(o_nd[0:real_size].asnumpy())
       613
       614         1            1      1.0      0.0              if return_data:
       615                                                           for j, x in enumerate(batch.data):
       616                                                               data_list[j].append(x[0:real_size].asnumpy())
       617                                                           for j, x in enumerate(batch.label):
       618                                                               label_list[j].append(x[0:real_size].asnumpy())
       619
       620         2           44     22.0      0.0          outputs = [np.concatenate(x) for x in output_list]
       621         1            2      2.0      0.0          if len(outputs) == 1:
       622         1            1      1.0      0.0              outputs = outputs[0]
       623
       624         1            1      1.0      0.0          if return_data:
       625                                                       data = [np.concatenate(x) for x in data_list]
       626                                                       label = [np.concatenate(x) for x in label_list]
       627                                                       if len(data) == 1:
       628                                                           data = data[0]
       629                                                       if len(label) == 1:
       630                                                           label = label[0]
       631                                                       return outputs, data, label
       632                                                   else:
       633         1            1      1.0      0.0              return outputs

**profile _init_predictor()**

    Line #      Hits         Time  Per Hit   % Time  Line Contents
    ==============================================================
       516                                               @profile
       517                                               def _init_predictor(self, input_shapes):
       518                                                   """Initialize the predictor module for running prediction."""
       519         1            2      2.0      0.0          if self._pred_exec is not None:
       520                                                       return
       521                                                   # for now only use the first device
       522         1     43947371 43947371.0    100.0          pred_exec = self.symbol.simple_bind(self.ctx[0], grad_req='null', **dict(input_shapes))
       523         1         5249   5249.0      0.0          pred_exec.copy_params_from(self.arg_params, self.aux_params)
       524
       525         1          636    636.0      0.0          _check_arguments(self.symbol)
       526         1            1      1.0      0.0          self._pred_exec = pred_exec

**Comment by piiswrong** [comment]
If you build with gpu enabled opencv the first gpu memory allocation can be really slow.
All the subsequent ones should be fine.

The temporary solution is to build without opencv or with gpu disabled opencv.

**Comment by ufoym** [comment]
@piiswrong Problem fixed. Thanks!

**Comment by ysh329** [comment]
Thanks a lot!

**Comment by jmacglashan** [comment]
@piiswrong You said not building with GPU enabled opencv is a _temporary_ solution; is there an improvement being planned in the future? GPU-enabled open-cv is useful for my setting and myself and my colleagues also use docker containers which means the "first time" can be rather frequent, making this very slow startup somewhat painful.
